### PR TITLE
Add attachment support

### PR DESCRIPTION
**Pull request by Ghouli**
Breaks compatibility with the current version of Bamboo, but has been tested working with changes in this pull request: https://github.com/thoughtbot/bamboo/pull/156

### CODE REVIEW

**Comment by tgautier** [comment]
We're still waiting for bamboo to merge the PR

**Comment by aseigo** [comment]
Was just about to fork bamboo_smtp to finish up a branch that implements this, only to notice that it is already here :) Woo! I see that bamboo 1.0.0-rc1 is out with Attachment support .. any chance this will get merged in the near future? (Waiting on this for a project myself, and it came up in a thread on elixirforum.com last week too :)

Thanks for the adaptor btw.. great stuff :)

**Comment by tgautier** [comment]
Thanks @aseigo for the ping and of course @Ghouli for all the work 😉 

Could you please just rebase all your commits to only one so we can merge this with a _clean_ commit ?

File: lib/bamboo/adapters/smtp_adapter.ex
Comment (kdisneur) [inline lib/bamboo/adapters/smtp_adapter.ex]: What do you think of using the `add_smtp_line` function instead of doing the concatenation yourself?

```elixir
body
|> add_smtp_line("Content-Type: attachment.content_type; name=\"#{attachment.filename}\"")
|> add_smtp_line("Content-Disposition: attachment; filename=\"#{attachment.filename}\")
# ....
```

File: lib/bamboo/adapters/smtp_adapter.ex
Comment (kdisneur) [inline lib/bamboo/adapters/smtp_adapter.ex]: 1. `body` already ends with a newline, do we need to add one blank line between the body and the attachment?
     I guess it's useful in order to separate the header from the body. In this case I would suggest to add a blank line (i.e: `|> add_smtp_line("")`) in the `add_attachment` function (between the header and body function calls).
    It would make it obvious we want an empty line between header and body. Moreover, it would be more consistent with what we do elsewhere
2. What do you think of using `add_smtp_line` instead of doing the concatenation yourself?

```elixir
add_smtp_line(body, Base.encode64(data))
```

File: lib/bamboo/adapters/smtp_adapter.ex
Comment (kdisneur) [inline lib/bamboo/adapters/smtp_adapter.ex]: What do you think of using pattern matching on function definition?

Something like: 

```elixir
defp add_attachment(nil, multi_part_mixed_delimiter), do: ""
defp add_attachment(attachment, multi_part_mixed_delimiter) do
  # ...
end
```

File: lib/bamboo/adapters/smtp_adapter.ex
Comment (kdisneur) [inline lib/bamboo/adapters/smtp_adapter.ex]: What do you think of using pattern matching on function definition?

Something like: 

```elixir
defp add_attachments(body, %Bamboo.Email{attachments: nil}, multi_part_mixed_delimiter), do: body
defp add_attachments(body, %Bamboo.Email{attachments: attachments}, multi_part_mixed_delimiter) do
  # ...
end
```

File: lib/bamboo/adapters/smtp_adapter.ex
Comment (kdisneur) [inline lib/bamboo/adapters/smtp_adapter.ex]: 1. Do we need an extra blank line between the body and the delimiter? (`body` already ends with an `\r\n`)
2. What do you think of using the `add_smtp_line` function instead of doing the concatenation yourself?

```elixir
defp add_multipart_mixed_split(body, delimiter) do
  add_smtp_line(body, "-#{delimiter}")
end
```

**Review by kdisneur — COMMENTED** [review]
I added some comments, but they are mostly about style/consistency on the project.

Thanks a lot for the PR ! 💯

File: lib/bamboo/adapters/smtp_adapter.ex
Comment (kdisneur) [reply lib/bamboo/adapters/smtp_adapter.ex]: Moreover, I guess some interpolation is missing on `attachment.content_type`?

Comment (Ghouli) [reply lib/bamboo/adapters/smtp_adapter.ex]: Sorry, my english fails here, do you mean that we should check if attachment.content_type is set before using? If so, we won't have to do that here. Bamboo.Attachment already sets content_type to "application/octet-stream" if not set on creation. If you meant something else, please let me now and I'll fix it. Thanks for the suggestions! :)

**Comment by Ghouli** [comment]
Hi, and thanks for the feedback. I've made those changes and pushed a commit. Now if I only knew how to make that rebase... I updated my fork from your codebase a few months ago and I'm worrying it will mess things up.

Need a little help here. So should I just pick the initial commit, squash all relevant ones after that, and pick those from your codebase which happened before I updated?

File: lib/bamboo/adapters/smtp_adapter.ex
Comment (kdisneur) [reply lib/bamboo/adapters/smtp_adapter.ex]: Sorry if my comment wasn't clear enough  :)

I meant we should use the value of the `attachment.content_type` variable instead of the literal string `"attachement.content_type"`:

```diff
- |> add_smtp_line("Content-Type: attachment.content_type; name=\"#{attachment.filename}\"")
+ |> add_smtp_line("Content-Type: #{attachment.content_type}; name=\"#{attachment.filename}\"")
```

**Comment by tgautier** [comment]
To rebase all your commits only to one and be sure you're up to date you can do the following.

1. rebase your current `develop` branch with ours
2. fix conflicts if any
3. `git rebase -i` your branch against our latest commit on develop, which is 0c296ad

Please let us know if you need any help

File: lib/bamboo/adapters/smtp_adapter.ex
Comment (kdisneur) [inline lib/bamboo/adapters/smtp_adapter.ex]: I replied to your question but as the discussion has been collapsed by GitHub, you probably missed it. So, I add it back here. :)

> Ghouli:
> Sorry, my english fails here, do you mean that we should check if attachment.content_type is set before using? If so, we won't have to do that here. Bamboo.Attachment already sets content_type to "application/octet-stream" if not set on creation. If you meant something else, please let me now and I'll fix it. Thanks for the suggestions! :)

Sorry if my comment wasn't clear enough :)

I meant we should use the value of the attachment.content_type variable instead of the literal string "attachement.content_type":

```diff
- |> add_smtp_line("Content-Type: attachment.content_type; name=\"#{attachment.filename}\"")
+ |> add_smtp_line("Content-Type: #{attachment.content_type}; name=\"#{attachment..filename}\"")
```

File: lib/bamboo/adapters/smtp_adapter.ex
Comment (kdisneur) [inline lib/bamboo/adapters/smtp_adapter.ex]: I just checkout your branch to test if it works but we can't send emails because `data` doesn't exist on the `%Bamboo.Attachment{}` struct anymore. 
I looked at one of the official adapters and they use `File.read!(attachment.path)` to get the data.

I would suggest to do the same here :)

**Review by kdisneur — CHANGES_REQUESTED** [review]

**Comment by kdisneur** [comment]
After fixing locally the error on `attachment.data`:

```diff
diff --git a/lib/bamboo/adapters/smtp_adapter.ex b/lib/bamboo/adapters/smtp_adapter.ex
index 7758090..4dde0e9 100644
--- a/lib/bamboo/adapters/smtp_adapter.ex
+++ b/lib/bamboo/adapters/smtp_adapter.ex
@@ -192,7 +192,7 @@ defmodule Bamboo.SMTPAdapter do
     |> add_multipart_mixed_split(multi_part_mixed_delimiter)
     |> add_attachment_header(attachment)
     |> add_smtp_line("")
-    |> add_attachment_body(attachment.data)
+    |> add_attachment_body(File.read!(attachment.path))
   end

   defp add_attachments(body, %Bamboo.Email{attachments: nil}, _), do: body
```

I did some tests but I can't manage to send emails.

Here my result

File: lib/bamboo/adapters/smtp_adapter.ex
Comment (Ghouli) [reply lib/bamboo/adapters/smtp_adapter.ex]: It doesn't exist _yet_ :) This was done as a preparation for this pull request on Bamboo: https://github.com/thoughtbot/bamboo/pull/292

More discussion here: https://github.com/thoughtbot/bamboo/issues/286

**Comment by Ghouli** [comment]
There were some issues regarding multipart delimiters which messed up emails, those are fixed now. Still waiting for this https://github.com/thoughtbot/bamboo/pull/292 to get merged and doing necessary changes after that.

**Comment by kdisneur** [comment]
@Ghouli Thanks, just tried it and it works well. 💯 

Hopefully https://github.com/thoughtbot/bamboo/pull/292 will be merged soon :)

**Comment by gitviola** [comment]
@Ghouli @kdisneur we just merged thoughtbot/bamboo#292 😊

**Comment by tgautier** [comment]
@schurig Thanks for letting us know!

We will merge this and prepare a new release soon.

**Comment by Ghouli** [comment]
Should probably wait for new release of bamboo before merging? This currently points to @schurig 's fork.

**Comment by tgautier** [comment]
thoughtbot/bamboo#156 and thoughtbot/bamboo#292 are merged, we will wait for a new release but we shouldn't have to wait for too long since they had a release candidate mainly for attachments support.

**Comment by ssnickolay** [comment]
IMHO this is not very clear solution, because I tried to use

```elixir
{:bamboo, "~> 1.0.0-rc.1"},
{:bamboo_smtp, git: "https://github.com/ssnickolay/bamboo_smtp.git", branch: "develop"}, # bamboo_smtp.develop with code form current PR.
```

and wrote the similar example

```elixir
defmodule MyController do
  use Krake.Web, :controller
  import Bamboo.Email

  def create(conn, params) do
    attachments = Enum.map(params["attachments"], &Bamboo.Attachment.new/1)
    new_email(
      to: "foo@example.com",
      from: "bar@example.com",
      subject: "subject",
      text_body: "body",
      attachments: attachments
    ) |> My.Mailer.deliver_now
  end
end
```

and got error like @kdisneur 

```elixir
** (KeyError) key :data not found in: %Bamboo.Attachment{content_type: "image/jpg", filename: "ring.jpg", path: "/Users/ssnickolay/Documents/bulk/ring.jpg"}
    (bamboo_smtp) lib/bamboo/adapters/smtp_adapter.ex:199: Bamboo.SMTPAdapter.add_attachment/2
```

I think that many users will get with the same problems...

For resolve this situation I implemented wrapper around original `Bamboo.Attachment` struct

```elixir
defmodule Bamboo.FileToolset do
  def load(%{path: path}), do: File.read!(path)
end

defmodule Bamboo.SMTPAttachment do
  alias Bamboo.Attachment
  defstruct filename: nil, content_type: nil, path: nil, data: nil

  @default_toolset Bamboo.FileToolset

  def new(path, opts \\ [])
  def new(path, [toolset: toolset]=opts) do
    attachment = Attachment.new(path, opts)
    _new(attachment, toolset.load(attachment))
  end

  def new(path, opts) do
    attachment = Attachment.new(path, opts)
    _new(attachment, @default_toolset.load(attachment))
  end

  defp _new(attachment, data) do
    %Attachment{filename: filename, content_type: content_type, path: path} = attachment
    %__MODULE__{filename: filename, content_type: content_type, path: path, data: data}
  end
end
```

so, you can use it like in original example

```elixir
def create(conn, params) do
  attachments = Enum.map(params["attachments"], &Bamboo.SMTPAttachment.new/1)
  ...
end
```

and you can implement custom toolsets for conversation `Bamboo.Attachment` to data.
What do you think @Ghouli ? I can make PR to your develop branch

**Comment by gitviola** [comment]
@ssnickolay in case you want to improve the attachment api of Bamboo it's not too late yet. They welcome everyone who is looking for improvements!

**Comment by Ghouli** [comment]
This doesn't work with bamboo 1.0.0-rc.1. Pull request adding `data` to `Bamboo.Attachment` was merged to master week ago, still waiting for updated release.

**Comment by ssnickolay** [comment]
@Ghouli Damn! >_< Sorry...
@schurig thx, I will do pr with "toolset  improvements" :)

**Comment by teehuhta** [comment]
This should probably be ok now?

**Comment by knewter** [comment]
@tgautier how's this PR at present? Anything I can help with to get it over the finish line? Going to use a fork for now but I'd love to help get support upstreamed.

**Comment by Ghouli** [comment]
This works with current master branch of bamboo in GitHub, but not with the one in hex.pm. They should probably up the version in mix.

**Comment by chungwong** [comment]
Does anyone have an update on this?
```
[
  {:bamboo, "~> 1.0.0-rc.2"},
  {:bamboo_smtp, "~> 1.5.0-rc.1"}
]
```
With 
```
email
|> put_attachment(params["file"])
```
Still throws `the Elixir.Bamboo.SMTPAdapter does not support attachments yet.`

**Comment by Ghouli** [comment]
This now works with bamboo 1.0.0-rc.2 in hex.pm.

**Comment by shivno** [comment]
i tested this and got: `#5.0.0 smtp; 500 Line too long>`
so the attachment has to be split with newlines, all other mails seem to
break the line at 76 chars. perhaps with a stream to save memory,
i dont know how to convert a string to a stream though. will try later.

**Comment by Ghouli** [comment]
Good catch. Should be [fixed](https://github.com/Ghouli/bamboo_smtp/commit/387beb02badf61b8f13100bec9d4c61fa0fb4fbf#diff-3445cccd639480a1c14d201eadfd12d3R190) now.

**Comment by tgautier** [comment]
Hey, can you please rebase this branch with develop so that we can release a new rc version for `bamboo_smtp` using the latest version `1.0.0-rc.2` for `bamboo`?

Thanks a lot.

**Comment by Ghouli** [comment]
Done.

**Comment by tgautier** [comment]
Thanks a lot, I'm waiting for @kdisneur to approve and it's a go.

**Comment by GalliGit** [comment]
Hi, can i put attachment already with bamboo_smtp v1.5.0-rc.1 or i have to wait for an update? How to put a single attachment ?

**Review by kdisneur — APPROVED** [review]

**Comment by kdisneur** [comment]
Thank you all for all the work you did on this branch! 🙇 

We are going to create a new RC release.

**Pull request merged**
